master/pkg/searcher/grid.go

```diff
@@ -106,36 +106,94 @@ func (s *gridSearch) trialClosed(ctx context, _ model.RequestID) ([]Operation, e
 }
 
 func newHyperparameterGrid(params expconf.Hyperparameters) []hparamSample {
-	var names []string
-	var values [][]interface{}
+	hpToInd, valueSets := getHPsValueSets(params)
+	values := cartesianProduct(valueSets)
+	var samples []hparamSample
+	for _, val := range values {
+		samples = append(samples, createHparamSample(params, hpToInd, val))
+	}
+	return samples
+}
+
+// Returns gridded values for a single hyperparameter.
+func getHPValueSet(h expconf.Hyperparameter) [][]interface{} {
+	var valueSets [][]interface{}
+	if h.RawNestedHyperparameter != nil {
+		for _, nestedHP := range *h.RawNestedHyperparameter {
+			valueSets = append(valueSets, getHPValueSet(nestedHP)...)
+		}
```

Comment:
is it a problem that this ordering is non-deterministic?

I *think* so, because it seems like the recursion is happening in two separate places, but I can't tell.

Either way, probably sorting the keys is a good idea to make the code safe for future reuse.

Reply:
We won't have issues with nondeterministic ordering because we store the `valueSet` index for a given hyperparameter in `hpToInd` and use that to map values to a nested `hparamSample`.